contracts/token/ERC20/extensions/draft-ERC20TemporaryApproval.sol

```diff
@@ -0,0 +1,106 @@
+ *
+ * WARNING: This is a draft contract. The corresponding ERC is still subject to changes.
+ */
+abstract contract ERC20TemporaryApproval is ERC20, IERC7674 {
+    using SlotDerivation for bytes32;
+    using StorageSlot for bytes32;
+    using StorageSlot for StorageSlot.Uint256SlotType;
+
+    // keccak256(abi.encode(uint256(keccak256("openzeppelin.storage.ERC20TemporaryApproval")) - 1)) & ~bytes32(uint256(0xff))
+    // solhint-disable-next-line const-name-snakecase
+    bytes32 private constant ERC20TemporaryApprovalStorageLocation =
+        0x0fd66af0be6cb88466bb5c49c7ea8fbb4acdc82057e863d0a17fddeaaf18fe00;
+
+    /**
+     * @dev {allowance} override that includes the temporary allowance when looking up the current allowance. If
+     * adding up the persistent and the temporary allowances result in an overflow, type(uint256).max is returned.
+     */
+    function allowance(address owner, address spender) public view virtual override returns (uint256) {
+        (bool success, uint256 amount) = Math.tryAdd(
+            super.allowance(owner, spender),
+            _temporaryAllowance(owner, spender)
+        );
+        return success ? amount : type(uint256).max;
+    }
+
+    /**
+     * @dev Internal getter for the current temporary allowance that `spender` has over `owner` tokens.
+     */
+    function _temporaryAllowance(address owner, address spender) internal view virtual returns (uint256) {
+        return ERC20TemporaryApprovalStorageLocation.deriveMapping(owner).deriveMapping(spender).asUint256().tload();
+    }
+
+    /**
+     * @dev Alternative to {approve} that sets a `value` amount of tokens as the temporary allowance of `spender` over
+     * the caller's tokens.
+     *
+     * Returns a boolean value indicating whether the operation succeeded.
+     *
+     * Requirements:
+     * - `spender` cannot be the zero address.
+     *
+     * Does NOT emit an {Approval} event.
+     */
+    function temporaryApprove(address spender, uint256 value) public virtual returns (bool) {
+        _temporaryApprove(_msgSender(), spender, value);
+        return true;
+    }
+
+    /**
+     * @dev Sets `value` as the temporary allowance of `spender` over the `owner` s tokens.
+     *
+     * This internal function is equivalent to `temporaryApprove`, and can be used to e.g. set automatic allowances
+     * for certain subsystems, etc.
+     *
+     * Requirements:
+     * - `owner` cannot be the zero address.
+     * - `spender` cannot be the zero address.
+     *
+     * Does NOT emit an {Approval} event.
+     */
+    function _temporaryApprove(address owner, address spender, uint256 value) internal virtual {
+        if (owner == address(0)) {
+            revert ERC20InvalidApprover(address(0));
+        }
+        if (spender == address(0)) {
+            revert ERC20InvalidSpender(address(0));
+        }
+        ERC20TemporaryApprovalStorageLocation.deriveMapping(owner).deriveMapping(spender).asUint256().tstore(value);
```

Comment:
i think we can save a mload and keccak with something like:
```
  mstore(0x20, owner)
  mstore(0x0c, ERC20TemporaryApprovalStorageLocation)
  mstore(0x00, spender)
  tstore(keccak256(0x0c, 0x34), value)
```

Reply:
this would also work for `_temporaryAllowance`

Reply:
What I'm proposing matches the derivation of solidity. That means in the future, it solidity supports it, we can move to 
```solidity
struct ERC20TemporaryApprovalStorage {
    mapping(address account => mapping(address spender => uint256)) private _tAllowances;
}

function _getERC20TemporaryApprovalStorage() private pure returns (ERC20TemporaryApprovalStorage transient $) {
    assembly {
        $.slot := ERC20TemporaryApprovalStorageLocation
    }
}
```
and that would keep using the same locations

Reply:
Also:
- the `deriveMapping` function use the scratch space. They do not expand memory (cost is ~50gas each time)
- doing a single hash operation will require having all 3 values next to each other. both addresses are 20 bytes, to have that fit into the scratch space, we would need the "prefix" (mapping location equivalent) to be 24 bytes and not 32 as usual. If we stick with the usual 32 bytes, we cannot use the scratch space and are potentially extending memory, which has a cost.